content/en/blog/_posts/2023-04-11-topology-spread-features.md

```diff
@@ -0,0 +1,140 @@
+and newly created replicaset has the following `topologySpreadConstraints` in template.
+
+```yaml
+topologySpreadConstraints:
+- maxSkew: 1
+  minDomains: 5 # requires 5 Nodes at least.
+  whenUnsatisfiable: DoNotSchedule # minDomains is valid only when DoNotSchedule is used.
+  topologyKey: kubernetes.io/hostname 
+  labelSelector:
+    matchLabels:
+        foo: bar
+```
+
+This case, 3 Pods will be scheduled to those 3 Nodes,
+but other 2 Pods from this replicaset will be unschedulable until more Nodes join the cluster.
+
+The cluster autoscaler provisions new Nodes based on these unschedulable Pods,
+and as a result, the replicas are finally spread over 5 Nodes.
+
+## KEP-3094: Take taints/tolerations into consideration when calculating PodTopologySpread skew
+
+Before this, when we deploy a pod with `podTopologySpread` configured, we'll take all
+affinity nodes(satisfied with pod nodeAffinity and nodeSelector) into consideration
+in filtering and scoring, but a node with pod untolerated taint may also be a candidate
+because we didn't take care of node taints, which will lead to the pod pending.
+
+To avoid this and make a more fine-gained decision in scheduling, we introduced two new fields in
+`TopologySpreadConstraint` to define node inclusion policies including nodeAffinity and nodeTaint.
+
+It mostly looks like:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  # Configure a topology spread constraint
+  topologySpreadConstraints:
+    - maxSkew: <integer>
+      # ...
+      nodeAffinityPolicy: [Honor|Ignore]
+      nodeTaintsPolicy: [Honor|Ignore]
+  # other Pod fields go here
+```
+
+**nodeAffinityPolicy** indicates how we'll treat Pod's nodeAffinity/nodeSelector in pod topology spreading.
+If `Honor`, we'll filter out nodes not matching nodeAffinity/nodeSelector in calculation.
+If `Ignore`, these nodes will be included instead.
+
+For backwards-compatibility, nodeAffinityPolicy is default to `Honor`.
+
+**nodeTaintsPolicy** indicates how we'll treat node taints in pod topology spreading.
+If `Honor`, only tainted nodes for which the incoming pod has a toleration, will be included in calculation.
+If `Ignore`, we'll not consider the node taints at all in calculation, so a node with pod untolerated taint
+will also be included.
+
+For backwards-compatibility, nodeTaintsPolicy is default to the `Ignore`.
+
+The feature was introduced in v1.25 as alpha level. By default, it was disabled, so if you want to use this feature in v1.25,
+you have to enable the feature gate `NodeInclusionPolicyInPodTopologySpread` actively. In the following v1.26, we graduated
+this feature to beta and it was enabled by default since.
+
+## KEP-3243: Respect PodTopologySpread after rolling upgrades
+
+Pod Topology Spread uses the fields `topologyKey` or `labelSelector` to identify the group of pods over which
+spreading will be calculated. But it applies to all pods in a Deployment irrespective of their owning
```

Comment:
This is inaccurate. It depends on which label selector you use.

Maybe you can start like this:
When using topology spreading with Deployments, it is common practice to use the labelSelector of the Deployment as the label selector in the topology spread constraints. However, this implies that all pods of a Deployment are part of the spreading calculation, regardless of whether they belong to different revisions.

Reply:
Done